libnavigation-core/src/main/java/com/mapbox/navigation/core/routealternatives/RouteAlternativesController.kt

```diff
@@ -0,0 +1,115 @@
+    private val navigator: MapboxNativeNavigator,
+    private val directionsSession: DirectionsSession,
+    private val tripSession: TripSession,
+    private val routeOptionsUpdater: RouteOptionsUpdater
+) {
+    private val jobControl = ThreadController.getMainScopeAndRootJob()
+
+    private val mapboxTimer = MapboxTimer()
+    private val observers = CopyOnWriteArraySet<RouteAlternativesObserver>()
+    private var currentRequestId: Long? = null
+
+    fun register(routeAlternativesObserver: RouteAlternativesObserver) {
+        val needsToStartTimer = observers.isEmpty()
+        observers.add(routeAlternativesObserver)
+        if (needsToStartTimer) {
+            mapboxTimer.restartAfterMillis = options.intervalMillis
+            mapboxTimer.startTimer {
+                requestRouteAlternatives()
+            }
+        }
+    }
+
+    fun unregister(routeAlternativesObserver: RouteAlternativesObserver) {
+        observers.remove(routeAlternativesObserver)
+        if (observers.isEmpty()) {
+            stop()
+        }
+    }
+
+    fun unregisterAll() {
+        observers.clear()
+        stop()
+    }
+
+    private fun stop() {
+        mapboxTimer.stopJobs()
+        jobControl.job.cancelChildren()
+        currentRequestId?.let { directionsSession.cancelRouteRequest(it) }
+    }
+
+    private fun requestRouteAlternatives() {
+        if (directionsSession.routes.isEmpty() ||
+            tripSession.getState() != TripSessionState.STARTED
+        ) {
+            return
+        }
+
+        val routeOptionsResult = routeOptionsUpdater.update(
+            directionsSession.getPrimaryRouteOptions(),
+            tripSession.getRouteProgress(),
+            tripSession.getEnhancedLocation()
+        )
+
+        when (routeOptionsResult) {
+            is RouteOptionsUpdater.RouteOptionsResult.Success -> {
+                currentRequestId?.let { directionsSession.cancelRouteRequest(it) }
+                currentRequestId = directionsSession.requestRoutes(
+                    routeOptionsResult.routeOptions,
+                    routesRequestCallback
+                )
+            }
+            is RouteOptionsUpdater.RouteOptionsResult.Error -> {
+                logger.e(
+                    msg = Message("Route alternatives options are not available"),
+                    tr = routeOptionsResult.error
+                )
+            }
+        }
+    }
+
+    private val routesRequestCallback = object : RoutesRequestCallback {
+        override fun onRoutesReady(routes: List<DirectionsRoute>) {
+            val routeProgress = tripSession.getRouteProgress()
+                ?: return
+            jobControl.scope.launch {
+                if (tripSession.getState() == TripSessionState.STARTED) {
+                    val alternatives = routes.filter { navigator.isDifferentRoute(it) }
+                    observers.forEach { it.onRouteAlternatives(routeProgress, alternatives) }
```

Comment:
We also need to check if the route hasn't changed while the request was in flight to avoid a bug like https://github.com/mapbox/mapbox-navigation-android/issues/4409.

Reply:
yeah thanks for cutting a ticket. that issue is in a few places

Reply:
As for this fix. https://github.com/mapbox/mapbox-navigation-android/pull/4375/files#diff-9d5de4542e0fc70f55cc5dc3007a2636f073fb9a262f37b906bfcae09b72140eR142

When the routes are set it restarts the interval. Thinking about times when a new route is set and the interval happens to be close to requesting, would cause an inconsistent callback.